# Solved Lab | Data Structuring and Combining Data

---
## Solved Challenge 1: Combining & Cleaning Data

In this challenge, we will be working with the customer data from an insurance company, as we did in the two previous labs. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

But this time, we got new data, which can be found in the following 2 CSV files located at the links below.

- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv

Note that you'll need to clean and format the new data.

Observation:
- One option is to first combine the three datasets and then apply the cleaning function to the new combined dataset
- Another option would be to read the clean file you saved in the previous lab, and just clean the two new files and concatenate the three clean datasets

### OPTION ONE: combine the three datasets and then apply the cleaning function to the new combined dataset

In [1012]:
import pandas as pd

# read the three csv files into seperate dataframes
df1 = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv')
df2 = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv')
df3 = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv')

# display the 1st dataframes
print("DataFrame 1:")
print("Shape of df1:", df1.shape)
display(df1.head())
print(*80*"-")

# display the 2nd dataframes
print("DataFrame 2:")
print("Shape of df2:", df2.shape)
display(df2.head())
print(*80*"-")

# display the 3rd dataframes
print("DataFrame 3:")
print("Shape of df3:", df3.shape)
display(df3.head())
print(*80*"-")

# standardize column names in all three dataframes to the same column names
# 'State' to 'ST' in df3
df3 = df3.rename(columns={'State': 'ST'})
# 'Gender' to 'GENDER' in df3
df3 = df3.rename(columns={'Gender': 'GENDER'})

    
# ------------------------------------------------------------------------------------------------------------------------------------------------
# concatenate the three csv files into one dataframe
df_main = pd.concat([df1, df2, df3], ignore_index=True)

# display the main dataframe
print("Main DataFrame after concatenation:")
print("Shape of df_main:", df_main.shape)
display(df_main.head())

# save the main dataframe to a csv file
df_main.to_csv('combined_00_concat.csv', index=False)


DataFrame 1:
Shape of df1: (4008, 11)


Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    Income  Monthly Premium Auto Number of Open Complaints     Policy Type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   Vehicle Class  Total Claim Amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
DataFrame 2:
Shape of df2: (996, 11)


Customer          ST GENDER Education Customer Lifetime Value  Income  \
0  GS98873     Arizona      F  Bachelor              323912.47%   16061   
1  CW49887  California      F    Master              462680.11%   79487   
2  MY31220  California      F   College              899704.02%   54230   
3  UH35128      Oregon      F   College             2580706.30%   71210   
4  WH52799     Arizona      F   College              380812.21%   94903   

   Monthly Premium Auto Number of Open Complaints  Total Claim Amount  \
0                    88                    1/0/00               633.6   
1                   114                    1/0/00               547.2   
2                   112                    1/0/00               537.6   
3                   214                    1/1/00              1027.2   
4                    94                    1/0/00               451.2   

      Policy Type  Vehicle Class  
0   Personal Auto  Four-Door Car  
1    Special Auto            SUV  
2   Personal Auto   Two-Door Car  
3   Personal Auto     Luxury Car  
4  Corporate Auto   Two-Door Car

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
DataFrame 3:
Shape of df3: (7070, 11)


Customer       State  Customer Lifetime Value             Education Gender  \
0  SA25987  Washington              3479.137523  High School or Below      M   
1  TB86706     Arizona              2502.637401                Master      M   
2  ZL73902      Nevada              3265.156348              Bachelor      F   
3  KX23516  California              4455.843406  High School or Below      F   
4  FN77294  California              7704.958480  High School or Below      M   

   Income  Monthly Premium Auto  Number of Open Complaints    Policy Type  \
0       0                   104                          0  Personal Auto   
1       0                    66                          0  Personal Auto   
2   25820                    82                          0  Personal Auto   
3       0                   121                          0  Personal Auto   
4   30366                   101                          2  Personal Auto   

   Total Claim Amount  Vehicle Class  
0          499.200000   Two-Door Car  
1            3.468912   Two-Door Car  
2          393.600000  Four-Door Car  
3          699.615192            SUV  
4          484.800000            SUV

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Main DataFrame after concatenation:
Shape of df_main: (12074, 11)


Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    Income  Monthly Premium Auto Number of Open Complaints     Policy Type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   Vehicle Class  Total Claim Amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

<hr>

# 🧫 DATA CLEANING 🧫

<style>
h1 {
    text-align: center;
    color: blue;
    font-weight: bold;
}
</style>

<style>
h2 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h3 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h4 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<hr>

<hr>

## 0️⃣ DATA EXPLORATION


<style>
h1 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h2 {
    text-align: left;
    color: darkblue;
    font-weight: bold;
}
</style>

<style>
h3 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h4 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>
<hr>

In [1013]:
import pandas as pd
from IPython.display import display

# Load online_shoppers_intention.csv into a df_main DataFrame
df = pd.read_csv("combined_00_concat.csv")

# Display df DataFrame SHAPE
print("exploring Dataset:")
print(f"Main Dataset Shape: [{df.shape[0]}] ROWS and [{df.shape[1]}] COLUMNS\n")

# Display df DataFrame HEAD
display(df.head(10))

# Display data types & missing values of each column in df DataFrame
print("Data Types & Missing Values of Each Column:")
display(df.info())

exploring Dataset:
Main Dataset Shape: [12074] ROWS and [11] COLUMNS



Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   
5  OC83172      Oregon      F              Bachelor              825629.78%   
6  XZ87318      Oregon      F               College              538089.86%   
7  CF85061     Arizona      M                Master              721610.03%   
8  DY87989      Oregon      M              Bachelor             2412750.40%   
9  BQ94931      Oregon      F               College              738817.81%   

    Income  Monthly Premium Auto Number of Open Complaints     Policy Type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   
5  62902.0                  69.0                    1/0/00   Personal Auto   
6  55350.0                  67.0                    1/0/00  Corporate Auto   
7      0.0                 101.0                    1/0/00  Corporate Auto   
8  14072.0                  71.0                    1/0/00  Corporate Auto   
9  28812.0                  93.0                    1/0/00    Special Auto   

   Vehicle Class  Total Claim Amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323  
5   Two-Door Car          159.383042  
6  Four-Door Car          321.600000  
7  Four-Door Car          363.029680  
8  Four-Door Car          511.200000  
9  Four-Door Car          425.527834

Data Types & Missing Values of Each Column:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12074 entries, 0 to 12073
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer                   9137 non-null   object 
 1   ST                         9137 non-null   object 
 2   GENDER                     9015 non-null   object 
 3   Education                  9137 non-null   object 
 4   Customer Lifetime Value    9130 non-null   object 
 5   Income                     9137 non-null   float64
 6   Monthly Premium Auto       9137 non-null   float64
 7   Number of Open Complaints  9137 non-null   object 
 8   Policy Type                9137 non-null   object 
 9   Vehicle Class              9137 non-null   object 
 10  Total Claim Amount         9137 non-null   float64
dtypes: float64(3), object(8)
memory usage: 1.0+ MB


None

In [1014]:
# Display the list of column names in df
print("Column Names in df_main:")
display(df.columns.tolist())

# Loop through each column and print unique values
for col in df.columns:
    unique_vals = df[col].unique()
    print(f"Column: {col}")
    print(f"Unique values ({len(unique_vals)}): {unique_vals}\n")

print(f"\nData types check:")
for col in df.columns:
    dtype = df[col].dtype
    unique_vals = df[col].nunique()
    print(f"  ➡️{col:<25} {str(dtype):<10} [{unique_vals}] unique values")

Column Names in df_main:


['Customer',
 'ST',
 'GENDER',
 'Education',
 'Customer Lifetime Value',
 'Income',
 'Monthly Premium Auto',
 'Number of Open Complaints',
 'Policy Type',
 'Vehicle Class',
 'Total Claim Amount']

Column: Customer
Unique values (9057): ['RB50392' 'QZ44356' 'AI49188' ... 'TD14365' 'UP19263' 'Y167826']

Column: ST
Unique values (9): ['Washington' 'Arizona' 'Nevada' 'California' 'Oregon' 'Cali' 'AZ' 'WA'
 nan]

Column: GENDER
Unique values (6): [nan 'F' 'M' 'Femal' 'Male' 'female']

Column: Education
Unique values (7): ['Master' 'Bachelor' 'High School or Below' 'College' 'Bachelors' 'Doctor'
 nan]

Column: Customer Lifetime Value
Unique values (8212): [nan '697953.59%' '1288743.17%' ... '8163.890428' '7524.442436'
 '2611.836866']

Column: Income
Unique values (5656): [    0. 48767. 36357. ... 66367. 71941. 21941.]

Column: Monthly Premium Auto
Unique values (210): [ 1000.    94.   108.   106.    68.    69.    67.   101.    71.    93.
   110.    70.    64.    72.    74.    79.    73.    61.   111.    80.
    63.   237.    65.   130.    62.   117.    91.    90.    66.    99.
   103. 35354.   172.    92.   104.   113.   124.   105.   100.   112.
    97.    95.   102.    96.   186.   

<hr>

## 1️⃣ COLUMNS NAMES STANDARDIZING


<style>
h1 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h2 {
    text-align: left;
    color: darkblue;
    font-weight: bold;
}
</style>

<style>
h3 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h4 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>
<hr>

In [1015]:
import pandas as pd

# Define the Renaming Dictionary
rename_columns = {
    'Customer': 'customer_id',
    'ST': 'state',
    'GENDER': 'gender',
    'Education': 'education',
    'Customer Lifetime Value': 'customer_lifetime_value',
    'Income': 'income',
    'Monthly Premium Auto': 'monthly_premium_auto',
    'Number of Open Complaints': 'number_open_complaints',
    'Policy Type': 'policy_type',
    'Vehicle Class': 'vehicle_class',
    'Total Claim Amount': 'total_claim_amount'
}


# Rename columns using the rename Dictionary
df.rename(columns=rename_columns, inplace=True)

# Save updated dataset into a new CSV file
df.to_csv("combined_01_standard.csv", index=False)

# Display Standardized DataFrame
print("Standardized DataFrame:")
display(df.head())

# Display the list of the standardidized column names in df
print("Standardized Column Names in df:")
display(df.columns.tolist())
display(df.info())

Standardized DataFrame:


customer_id       state gender             education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   

  customer_lifetime_value   income  monthly_premium_auto  \
0                     NaN      0.0                1000.0   
1              697953.59%      0.0                  94.0   
2             1288743.17%  48767.0                 108.0   
3              764586.18%      0.0                 106.0   
4              536307.65%  36357.0                  68.0   

  number_open_complaints     policy_type  vehicle_class  total_claim_amount  
0                 1/0/00   Personal Auto  Four-Door Car            2.704934  
1                 1/0/00   Personal Auto  Four-Door Car         1131.464935  
2                 1/0/00   Personal Auto   Two-Door Car          566.472247  
3                 1/0/00  Corporate Auto            SUV          529.881344  
4                 1/0/00   Personal Auto  Four-Door Car           17.269323

Standardized Column Names in df:


['customer_id',
 'state',
 'gender',
 'education',
 'customer_lifetime_value',
 'income',
 'monthly_premium_auto',
 'number_open_complaints',
 'policy_type',
 'vehicle_class',
 'total_claim_amount']

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12074 entries, 0 to 12073
Data columns (total 11 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   customer_id              9137 non-null   object 
 1   state                    9137 non-null   object 
 2   gender                   9015 non-null   object 
 3   education                9137 non-null   object 
 4   customer_lifetime_value  9130 non-null   object 
 5   income                   9137 non-null   float64
 6   monthly_premium_auto     9137 non-null   float64
 7   number_open_complaints   9137 non-null   object 
 8   policy_type              9137 non-null   object 
 9   vehicle_class            9137 non-null   object 
 10  total_claim_amount       9137 non-null   float64
dtypes: float64(3), object(8)
memory usage: 1.0+ MB


None

<hr>

## 2️⃣ INVALID VALUES


<style>
h1 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h2 {
    text-align: left;
    color: darkblue;
    font-weight: bold;
}
</style>

<style>
h3 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h4 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>
<hr>

In [1016]:
import pandas as pd
# Load the standardized dataset into a new DataFrame
df = pd.read_csv("combined_01_standard.csv")

# 01 - gender ---------------------------------------------------------------------------
print("Before: 'gender' unique values:", df['gender'].unique())
# mapping gender values
gender_mapping = {
    "female": "F",
    "Femal": "F",    
    "Male": "M"
}
# replace values in gender
df['gender'] = df['gender'].replace(gender_mapping)
# Check GENDER unique values after mapping
print("After: 'gender' unique values:", df['gender'].unique())
print(*70*"-")

# 02 - state ---------------------------------------------------------------------------
print("Before: 'state' unique values:", df['state'].unique())
# mapping gender values
state_mapping = {
    "WA": "Washington",
    "AZ": "Arizona",    
    "Cali": "California"
}
# replace values in state
df['state'] = df['state'].replace(state_mapping)
# Check STATE unique values after mapping
print("After: 'state' unique values:", df['state'].unique())
print(*70*"-")

# 03 - education ---------------------------------------------------------------------------
print("Before: 'education' unique values:", df['education'].unique())
# replace Bachelors with Bachelor in education_level
df['education'] = df['education'].replace('Bachelors', 'Bachelor')
# Check education unique values after mapping
print("After: 'education' unique values:", df['education'].unique())
print(*70*"-")

# 04 - customer_lifetime_value ---------------------------------------------------------------------------
print("Before: 'customer_lifetime_value' unique values:", df['customer_lifetime_value'].unique())
# Remove '%' and convert to float
df['customer_lifetime_value'] = df['customer_lifetime_value'].str.replace('%', '', regex=False).astype(float)
# Check education_level unique values after mapping
print("After: 'customer_lifetime_value' unique values:", df['customer_lifetime_value'].unique())
print(*70*"-")

# 05 - vehicle_class --------------------------------------------------------------------------
print("Before: 'vehicle_class' unique values:", df['vehicle_class'].unique())
# mapping vehicle_class values
vehicle_mapping = {
    "Sports Car": "Luxury",
    "Luxury SUV": "Luxury",
    "Luxury Car": "Luxury"
}
# replace values in education_level
df['vehicle_class'] = df['vehicle_class'].replace(vehicle_mapping)
# Check unique values after mapping
print("After: 'vehicle_class' unique values:", df['vehicle_class'].unique())
print(*70*"-")

# 06 - number_open_complaints --------------------------------------------------------------------------
print("Before: 'number_open_complaints' unique values:", df['number_open_complaints'].unique())
# Extract the number of open complaints from the string and convert to numeric
df['number_open_complaints'] = pd.to_numeric(df['number_open_complaints'].str.split('/', expand=True)[1],errors='coerce')
# Check unique values after mapping
print("After: 'number_open_complaints' unique values:", df['number_open_complaints'].unique())
print(*70*"-")

# Display the dataset after dealing with invalid values ----------------------------------------------------------------------
print("Dataset after dealing with invalid values:")
display(df.head())

# Save step 02 deal with invalid values dataset into a new CSV file ------------------------------
df.to_csv("combined_02_invalid-values.csv", index=False)


Before: 'gender' unique values: [nan 'F' 'M' 'Femal' 'Male' 'female']
After: 'gender' unique values: [nan 'F' 'M']
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Before: 'state' unique values: ['Washington' 'Arizona' 'Nevada' 'California' 'Oregon' 'Cali' 'AZ' 'WA'
 nan]
After: 'state' unique values: ['Washington' 'Arizona' 'Nevada' 'California' 'Oregon' nan]
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Before: 'education' unique values: ['Master' 'Bachelor' 'High School or Below' 'College' 'Bachelors' 'Doctor'
 nan]
After: 'education' unique values: ['Master' 'Bachelor' 'High School or Below' 'College' 'Doctor' nan]
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Before: 'customer_lifetime_value' unique values:

customer_id       state gender             education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   

   customer_lifetime_value   income  monthly_premium_auto  \
0                      NaN      0.0                1000.0   
1                697953.59      0.0                  94.0   
2               1288743.17  48767.0                 108.0   
3                764586.18      0.0                 106.0   
4                536307.65  36357.0                  68.0   

   number_open_complaints     policy_type  vehicle_class  total_claim_amount  
0                     0.0   Personal Auto  Four-Door Car            2.704934  
1                     0.0   Personal Auto  Four-Door Car         1131.464935  
2                     0.0   Personal Auto   Two-Door Car          566.472247  
3                     0.0  Corporate Auto            SUV          529.881344  
4                     0.0   Personal Auto  Four-Door Car           17.269323

<hr>

## 3️⃣ DATA TYPES


<style>
h1 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h2 {
    text-align: left;
    color: darkblue;
    font-weight: bold;
}
</style>

<style>
h3 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h4 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>
<hr>

In [1017]:
import pandas as pd

# Load the step 02 dataset into a new DataFrame
df = pd.read_csv("combined_02_invalid-values.csv")

# check data types before data formatting
print("Data types before data formatting:")
df.info()



Data types before data formatting:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12074 entries, 0 to 12073
Data columns (total 11 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   customer_id              9137 non-null   object 
 1   state                    9137 non-null   object 
 2   gender                   9015 non-null   object 
 3   education                9137 non-null   object 
 4   customer_lifetime_value  9130 non-null   float64
 5   income                   9137 non-null   float64
 6   monthly_premium_auto     9137 non-null   float64
 7   number_open_complaints   2067 non-null   float64
 8   policy_type              9137 non-null   object 
 9   vehicle_class            9137 non-null   object 
 10  total_claim_amount       9137 non-null   float64
dtypes: float64(5), object(6)
memory usage: 1.0+ MB


In [1018]:
# convert number_open_complaints to integer type
df['number_open_complaints'] = pd.to_numeric(df['number_open_complaints'], errors='coerce').astype('Int64')

# check data types after data formatting
print("Data types after data formatting:")
df.info()
print("*" * 70)

# display the dataset after data formatting ----------------------------------------------------------------------
print("Dataset after data formatting:")
print("Dataset shape:", df.shape)
display(df.head())

# save step 3 data formatting dataset into a new CSV file ------------------------------
df.to_csv("combined_03_data-formatting.csv", index=False)

Data types after data formatting:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12074 entries, 0 to 12073
Data columns (total 11 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   customer_id              9137 non-null   object 
 1   state                    9137 non-null   object 
 2   gender                   9015 non-null   object 
 3   education                9137 non-null   object 
 4   customer_lifetime_value  9130 non-null   float64
 5   income                   9137 non-null   float64
 6   monthly_premium_auto     9137 non-null   float64
 7   number_open_complaints   2067 non-null   Int64  
 8   policy_type              9137 non-null   object 
 9   vehicle_class            9137 non-null   object 
 10  total_claim_amount       9137 non-null   float64
dtypes: Int64(1), float64(4), object(6)
memory usage: 1.0+ MB
**********************************************************************
Dataset after data for

customer_id       state gender             education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   

   customer_lifetime_value   income  monthly_premium_auto  \
0                      NaN      0.0                1000.0   
1                697953.59      0.0                  94.0   
2               1288743.17  48767.0                 108.0   
3                764586.18      0.0                 106.0   
4                536307.65  36357.0                  68.0   

   number_open_complaints     policy_type  vehicle_class  total_claim_amount  
0                       0   Personal Auto  Four-Door Car            2.704934  
1                       0   Personal Auto  Four-Door Car         1131.464935  
2                       0   Personal Auto   Two-Door Car          566.472247  
3                       0  Corporate Auto            SUV          529.881344  
4                       0   Personal Auto  Four-Door Car           17.269323

<hr>

## 4️⃣ NULL VALUES


<style>
h1 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h2 {
    text-align: left;
    color: darkblue;
    font-weight: bold;
}
</style>

<style>
h3 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h4 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>
<hr>

In [1019]:
import pandas as pd
# read csv
df = pd.read_csv("combined_03_data-formatting.csv")

# Check for null values
null_counts = df.isnull().sum()
print(f"Total null count in the Dataframe = {null_counts.sum()}\n")

# Identify how many null values each column has. 
print("Columns with null values:")
display(null_counts[null_counts > 0])

print("-" * 70)

# show null values in each column
print("Null values in each column:")
# Loop through each column and print the number of nulls in each column
for col in df.columns:
    null_count = df[col].isnull().sum()
    print(f"Column '{col}'➡️[{null_count}] null values")

Total null count in the Dataframe = 39506

Columns with null values:


customer_id                 2937
state                       2937
gender                      3059
education                   2937
customer_lifetime_value     2944
income                      2937
monthly_premium_auto        2937
number_open_complaints     10007
policy_type                 2937
vehicle_class               2937
total_claim_amount          2937
dtype: int64

----------------------------------------------------------------------
Null values in each column:
Column 'customer_id'➡️[2937] null values
Column 'state'➡️[2937] null values
Column 'gender'➡️[3059] null values
Column 'education'➡️[2937] null values
Column 'customer_lifetime_value'➡️[2944] null values
Column 'income'➡️[2937] null values
Column 'monthly_premium_auto'➡️[2937] null values
Column 'number_open_complaints'➡️[10007] null values
Column 'policy_type'➡️[2937] null values
Column 'vehicle_class'➡️[2937] null values
Column 'total_claim_amount'➡️[2937] null values


In [1020]:
# show unique values of each column
print("\nUnique values in each column:")
for col in df.columns:
    unique_vals = df[col].unique()
    print(f"{col}")
    print(f"Unique values ({len(unique_vals)}): {unique_vals}\n")



Unique values in each column:
customer_id
Unique values (9057): ['RB50392' 'QZ44356' 'AI49188' ... 'TD14365' 'UP19263' 'Y167826']

state
Unique values (6): ['Washington' 'Arizona' 'Nevada' 'California' 'Oregon' nan]

gender
Unique values (3): [nan 'F' 'M']

education
Unique values (6): ['Master' 'Bachelor' 'High School or Below' 'College' 'Doctor' nan]

customer_lifetime_value
Unique values (8212): [           nan  697953.59     1288743.17     ...    8163.890428
    7524.442436    2611.836866]

income
Unique values (5656): [    0. 48767. 36357. ... 66367. 71941. 21941.]

monthly_premium_auto
Unique values (210): [ 1000.    94.   108.   106.    68.    69.    67.   101.    71.    93.
   110.    70.    64.    72.    74.    79.    73.    61.   111.    80.
    63.   237.    65.   130.    62.   117.    91.    90.    66.    99.
   103. 35354.   172.    92.   104.   113.   124.   105.   100.   112.
    97.    95.   102.    96.   186.    77.    76.   119.    87.   109.
   107.    88.    85.  6

In [1021]:
# null values in the dataset before dealing with null values
null_counts_before = df.isnull().sum()


# null values in education column BEFORE
print("BEFORE, null values in education column:", df['education'].isnull().sum())
# fill null values in education column with 'Not Specified'
df['education'] = df['education'].fillna('Not Specified')
# null values in education column AFTER
print("AFTER, null values in education column:", df['education'].isnull().sum())
print("-" * 70)

# null values in customer_lifetime_value BEFORE
print("BEFORE, null values in customer_lifetime_value column:", df['customer_lifetime_value'].isnull().sum())
# fill null values in customer_lifetime_value column with median value of the column
df['customer_lifetime_value'] = df['customer_lifetime_value'].fillna(df['customer_lifetime_value'].median())
# null values in customer_lifetime_value column AFTER
print("AFTER, null values in customer_lifetime_value column:", df['customer_lifetime_value'].isnull().sum()) 
print("-" * 70)

# null values in gender column
print("null values in gender column:", df['gender'].isnull().sum())
# fill null values in gender column with 'U' for unknown
df['gender'] = df['gender'].fillna('U')
# null values in gender column AFTER
print("AFTER, null values in gender column:", df['gender'].isnull().sum())
print("-" * 70)

# null values in state column
print("null values in state column:", df['state'].isnull().sum())
# fill null values in state column with 'Not Specified'
df['state'] = df['state'].fillna('Not Specified')
# null values in state column AFTER
print("AFTER, null values in state column:", df['state'].isnull().sum())
print("-" * 70)

# null values in number_open_complaints column
print("null values in number_open_complaints column:", df['number_open_complaints'].isnull().sum())
# fill number_open_complaints null values with 0
df['number_open_complaints'] = df['number_open_complaints'].fillna(0)
# null values in number_open_complaints column AFTER
print("AFTER, null values in number_open_complaints column:", df['number_open_complaints'].isnull().sum())
print("-" * 70)

# null values in vehicle_class column
print("null values in vehicle_class column:", df['vehicle_class'].isnull().sum())
# fill null values in vehicle_class column with 'Not Specified'
df['vehicle_class'] = df['vehicle_class'].fillna('Not Specified')
# null values in vehicle_class column AFTER
print("AFTER, null values in vehicle_class column:", df['vehicle_class'].isnull().sum())
print("-" * 70)

# null values in customer_id column
print("null values in customer_id column:", df['customer_id'].isnull().sum())
# drop rows with null values in customer_id column
df = df.dropna(subset=['customer_id'])
# null values in customer_id column AFTER
print("AFTER, null values in customer_id column:", df['customer_id'].isnull().sum())
print("-" * 70)

# after dealing with null values in the dataset, check for null values again
null_counts_after = df.isnull().sum()
print("Null values in each column after dealing with null values:")
display(null_counts_after)

# save in a new csv file the dataset after dealing with null values
df.to_csv("combined_04_no-nulls.csv", index=False)

# display the dataset after dealing with null values ----------------------------------------------------------------------
print("Dataset after dealing with null values:")
print("Dataset shape:", df.shape)
display(df.head())

BEFORE, null values in education column: 2937
AFTER, null values in education column: 0
----------------------------------------------------------------------
BEFORE, null values in customer_lifetime_value column: 2944
AFTER, null values in customer_lifetime_value column: 0
----------------------------------------------------------------------
null values in gender column: 3059
AFTER, null values in gender column: 0
----------------------------------------------------------------------
null values in state column: 2937
AFTER, null values in state column: 0
----------------------------------------------------------------------
null values in number_open_complaints column: 10007
AFTER, null values in number_open_complaints column: 0
----------------------------------------------------------------------
null values in vehicle_class column: 2937
AFTER, null values in vehicle_class column: 0
----------------------------------------------------------------------
null values in customer_id co

customer_id                0
state                      0
gender                     0
education                  0
customer_lifetime_value    0
income                     0
monthly_premium_auto       0
number_open_complaints     0
policy_type                0
vehicle_class              0
total_claim_amount         0
dtype: int64

Dataset after dealing with null values:
Dataset shape: (9137, 11)


customer_id       state gender             education  \
0     RB50392  Washington      U                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   

   customer_lifetime_value   income  monthly_premium_auto  \
0             7.714877e+03      0.0                1000.0   
1             6.979536e+05      0.0                  94.0   
2             1.288743e+06  48767.0                 108.0   
3             7.645862e+05      0.0                 106.0   
4             5.363077e+05  36357.0                  68.0   

   number_open_complaints     policy_type  vehicle_class  total_claim_amount  
0                     0.0   Personal Auto  Four-Door Car            2.704934  
1                     0.0   Personal Auto  Four-Door Car         1131.464935  
2                     0.0   Personal Auto   Two-Door Car          566.472247  
3                     0.0  Corporate Auto            SUV          529.881344  
4                     0.0   Personal Auto  Four-Door Car           17.269323

In [1022]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9137 entries, 0 to 12073
Data columns (total 11 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   customer_id              9137 non-null   object 
 1   state                    9137 non-null   object 
 2   gender                   9137 non-null   object 
 3   education                9137 non-null   object 
 4   customer_lifetime_value  9137 non-null   float64
 5   income                   9137 non-null   float64
 6   monthly_premium_auto     9137 non-null   float64
 7   number_open_complaints   9137 non-null   float64
 8   policy_type              9137 non-null   object 
 9   vehicle_class            9137 non-null   object 
 10  total_claim_amount       9137 non-null   float64
dtypes: float64(5), object(6)
memory usage: 856.6+ KB


<hr>

## 5️⃣ DUPLICATE VALUES


<style>
h1 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h2 {
    text-align: left;
    color: darkblue;
    font-weight: bold;
}
</style>

<style>
h3 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h4 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>
<hr>

In [1023]:
import pandas as pd

# read csv into a new DataFrame
df = pd.read_csv("combined_04_no-nulls.csv")

# Check for duplicate rows in the DataFrame
duplicate_rows = df[df.duplicated()]

# Shape of the dataset before dropping duplicate rows
print("Dataset shape before dropping duplicates:", df.shape)

# Print the number of duplicate rows
print(f"Number of duplicate rows: {duplicate_rows.shape[0]}")

# Display the duplicate rows
if duplicate_rows.empty:
    print("No duplicate rows found.")
else:
    print("\nDuplicate rows:")
    display(duplicate_rows)

print(*70*"-")

# drop duplicate rows, keeping the first occurrence
print("Dropping duplicate rows, keeping the first occurrence...")
df.drop_duplicates(keep='first', inplace=True)
duplicate_rows = df[df.duplicated()]

# Display the duplicate rows
if duplicate_rows.empty:
    print("No duplicate rows found.")
else:
    print("\nDuplicate rows:")
    display(duplicate_rows)
    
# save the dataset after dropping duplicate rows into a new CSV file
df.to_csv("combined_05_no-duplicates.csv", index=False)

Dataset shape before dropping duplicates: (9137, 11)
Number of duplicate rows: 3

Duplicate rows:


customer_id       state gender education  customer_lifetime_value  \
1071     GS98873     Arizona      F  Bachelor                323912.47   
1072     CW49887  California      F    Master                462680.11   
1073     MY31220  California      F   College                899704.02   

       income  monthly_premium_auto  number_open_complaints    policy_type  \
1071  16061.0                  88.0                     0.0  Personal Auto   
1072  79487.0                 114.0                     0.0   Special Auto   
1073  54230.0                 112.0                     0.0  Personal Auto   

      vehicle_class  total_claim_amount  
1071  Four-Door Car               633.6  
1072            SUV               547.2  
1073   Two-Door Car               537.6

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Dropping duplicate rows, keeping the first occurrence...
No duplicate rows found.


In [1024]:
import pandas as pd
# read csv into a new DataFrame
df = pd.read_csv("combined_05_no-duplicates.csv")

# display the dataset ----------------------------------------------------------------------
print("Clean Dataset:")
print("Clean Dataset shape:", df.shape)
display(df.head(20))

# save the clean dataset into a new CSV file
df.to_csv("combined_06_clean.csv", index=False)


Clean Dataset:
Clean Dataset shape: (9134, 11)


customer_id       state gender             education  \
0      RB50392  Washington      U                Master   
1      QZ44356     Arizona      F              Bachelor   
2      AI49188      Nevada      F              Bachelor   
3      WW63253  California      M              Bachelor   
4      GA49547  Washington      M  High School or Below   
5      OC83172      Oregon      F              Bachelor   
6      XZ87318      Oregon      F               College   
7      CF85061     Arizona      M                Master   
8      DY87989      Oregon      M              Bachelor   
9      BQ94931      Oregon      F               College   
10     SX51350  California      M               College   
11     VQ65197  California      U               College   
12     DP39365  California      U                Master   
13     SJ95423     Arizona      U  High School or Below   
14     IL66569  California      U               College   
15     BW63560      Oregon      U              Bachelor   
16     FV94802      Nevada      U  High School or Below   
17     OE15005  California      U               College   
18     WC83389      Oregon      U               College   
19     FL50705  California      U  High School or Below   

    customer_lifetime_value   income  monthly_premium_auto  \
0              7.714877e+03      0.0                1000.0   
1              6.979536e+05      0.0                  94.0   
2              1.288743e+06  48767.0                 108.0   
3              7.645862e+05      0.0                 106.0   
4              5.363077e+05  36357.0                  68.0   
5              8.256298e+05  62902.0                  69.0   
6              5.380899e+05  55350.0                  67.0   
7              7.216100e+05      0.0                 101.0   
8              2.412750e+06  14072.0                  71.0   
9              7.388178e+05  28812.0                  93.0   
10             4.738992e+05      0.0                  67.0   
11             8.197197e+05      0.0                 110.0   
12             8.798797e+05  77026.0                 110.0   
13             8.819019e+05  99845.0                 110.0   
14             5.384432e+05  83689.0                  70.0   
15             7.463139e+05  24599.0                  64.0   
16             2.566868e+05  25049.0                  67.0   
17             3.945242e+05  28855.0                 101.0   
18             5.710333e+05  51148.0                  72.0   
19             8.162617e+05  66140.0                 101.0   

    number_open_complaints     policy_type  vehicle_class  total_claim_amount  
0                      0.0   Personal Auto  Four-Door Car            2.704934  
1                      0.0   Personal Auto  Four-Door Car         1131.464935  
2                      0.0   Personal Auto   Two-Door Car          566.472247  
3                      0.0  Corporate Auto            SUV          529.881344  
4                      0.0   Personal Auto  Four-Door Car           17.269323  
5                      0.0   Personal Auto   Two-Door Car          159.383042  
6                      0.0  Corporate Auto  Four-Door Car          321.600000  
7                      0.0  Corporate Auto  Four-Door Car          363.029680  
8                      0.0  Corporate Auto  Four-Door Car          511.200000  
9                      0.0    Special Auto  Four-Door Car          425.527834  
10                     0.0   Personal Auto  Four-Door Car          482.400000  
11                     0.0   Personal Auto            SUV          528.000000  
12                     2.0  Corporate Auto  Four-Door Car          472.029737  
13                     1.0  Corporate Auto            SUV          528.000000  
14                     2.0  Corporate Auto  Four-Door Car          307.139132  
15                     1.0  Corporate Auto  Four-Door Car           42.920271  
16                     0.0   Personal Auto   Two-Door Car          454.245098  
17                  

---
## Challenge 2: Structuring Data

In this challenge, we will continue to work with customer data from an insurance company, but we will use a dataset with more columns, called `marketing_customer_analysis.csv`, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv

This dataset contains information such as `customer demographics`, `policy details`, `vehicle information`, and the `customer's response to the last marketing campaign`. 

Our **goal** is to **explore and analyze this data by performing data cleaning, formatting, and structuring**.

<hr>

## 0️⃣ DATA EXPLORATION


<style>
h1 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h2 {
    text-align: left;
    color: darkblue;
    font-weight: bold;
}
</style>

<style>
h3 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h4 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>
<hr>

In [1025]:
import pandas as pd

# read the three csv files into seperate dataframes
df = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv')


# Display df DataFrame
print("Exploring the Marketing Customer Dataset:")
# Display df DataFrame SHAPE
print(f"Main Dataset Shape: [{df.shape[0]}] ROWS and [{df.shape[1]}] COLUMNS\n")
# Display df DataFrame HEAD
display(df.head(30))

# Display data types & missing values of each column in df DataFrame
print("Data Types & Missing Values of Each Column:")
display(df.info())

# Check for null values
print(f"Total null count in the Dataframe = {null_counts.sum()}\n")
print("There are 0 null values in the dataset!")

Exploring the Marketing Customer Dataset:
Main Dataset Shape: [10910] ROWS and [27] COLUMNS



unnamed:_0 customer       state  customer_lifetime_value response  \
0            0  DK49336     Arizona              4809.216960       No   
1            1  KX64629  California              2228.525238       No   
2            2  LZ68649  Washington             14947.917300       No   
3            3  XL78013      Oregon             22332.439460      Yes   
4            4  QA50777      Oregon              9025.067525       No   
5            5  HO81102  California              4745.181764       No   
6            6  IW72280  California              5035.035257       No   
7            7  IH64929  California              4956.247235       No   
8            8  FM55990  California              5989.773931      Yes   
9            9  QX31376      Oregon             17114.328150       No   
10          10  HG93801     Arizona              5154.764074       No   
11          11  GE63585  California              2420.711888       No   
12          12  EO51920      Oregon             19955.257130       No   
13          13  KR82385  California              5454.587929       No   
14          14  TL59568     Arizona             12836.009510       No   
15          15  CW49887  California              4626.801093      Yes   
16          16  FH51383  California              5326.677654       No   
17          17  TR26615      Oregon              2497.531509       No   
18          18  ZC97968  California              2736.352663       No   
19          19  NJ54277  California              3746.751625      Yes   
20          20  OM80593  California              7247.313475       No   
21          21  RL65736      Oregon              6274.447705       No   
22          22  EQ62488      Oregon             15031.564500       No   
23          23  NQ71171  California              5107.071054       No   
24          24  YK32760      Oregon              2502.209885       No   
25          25  AY59197      Nevada              9521.222806       No   
26          26  XF15789     Arizona              8869.619888       No   
27          27  MQ68407      Oregon              4376.363592      Yes   
28          28  CI99444      Nevada              2926.886293       No   
29          29  BD70337  California             24706.920110       No   

    coverage             education effective_to_date employmentstatus gender  \
0      Basic               College        2011-02-18         Employed      M   
1      Basic               College        2011-01-18       Unemployed      F   
2      Basic              Bachelor        2011-02-10         Employed      M   
3   Extended               College        2011-01-11         Employed      M   
4    Premium              Bachelor        2011-01-17    Medical Leave      F   
5      Basic  High School or Below        2011-02-14         Employed      M   
6      Basic                Doctor        2011-02-14         Employed      F   
7      Basic  High School or Below        2011-02-24         Employed      M   
8    Premium               College        2011-01-19         Employed      M   
9   Extended                Master        2011-01-04         Employed      F   
10  Extended  High School or Below        2011-01-02         Employed      M   
11     Basic              Bachelor        2011-02-07         Employed      M   
12  Extended  High School or Below        2011-01-31         Employed      F   
13     Basic                Master        2011-01-26         Employed      M   
14   Premium  High School or Below        2011-02-28       Unemployed      M   
15     Basic                Master        2011-01-16         Employed      F   
16     Basic  High School or Below        2011-02-07         Employed      F   
17     Basic  High School or Below        2011-01-17       Unemployed      M   
18     Basic               College        2011-02-14         Disabled      M   
19  Extended               College        2011-02-26         Employed      F   
20     Basic              Bachelor        2011-02-23       U

Data Types & Missing Values of Each Column:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10910 entries, 0 to 10909
Data columns (total 27 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   unnamed:_0                     10910 non-null  int64  
 1   customer                       10910 non-null  object 
 2   state                          10910 non-null  object 
 3   customer_lifetime_value        10910 non-null  float64
 4   response                       10910 non-null  object 
 5   coverage                       10910 non-null  object 
 6   education                      10910 non-null  object 
 7   effective_to_date              10910 non-null  object 
 8   employmentstatus               10910 non-null  object 
 9   gender                         10910 non-null  object 
 10  income                         10910 non-null  int64  
 11  location_code                  10910 non-null  object 
 12  ma

None

Total null count in the Dataframe = 39506

There are 0 null values in the dataset!


<hr>

## 1️⃣ COLUMNS NAMES STANDARDIZING


<style>
h1 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h2 {
    text-align: left;
    color: darkblue;
    font-weight: bold;
}
</style>

<style>
h3 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h4 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>
<hr>

In [1026]:
import pandas as pd
from IPython.display import display

# drop the 'Unnamed: 0' column
df.drop(columns=['unnamed:_0'], inplace=True)

# save the dataset after dropping the 'unnamed:_0' column into a new CSV file
df.to_csv("marketing_customer_analysis_00_drop-column.csv", index=False)

# read csv into a new DataFrame
df = pd.read_csv("marketing_customer_analysis_00_drop-column.csv")

# Display df DataFrame
print("Exploring the Marketing Customer Dataset:")

# Display df DataFrame SHAPE
print(f"Main Dataset Shape: [{df.shape[0]}] ROWS and [{df.shape[1]}] COLUMNS\n")

# Display df DataFrame HEAD
display(df.head(30))

# Display data types & missing values of each column in df DataFrame
print("Data Types & Missing Values of Each Column:")
display(df.info())

Exploring the Marketing Customer Dataset:
Main Dataset Shape: [10910] ROWS and [26] COLUMNS



customer       state  customer_lifetime_value response  coverage  \
0   DK49336     Arizona              4809.216960       No     Basic   
1   KX64629  California              2228.525238       No     Basic   
2   LZ68649  Washington             14947.917300       No     Basic   
3   XL78013      Oregon             22332.439460      Yes  Extended   
4   QA50777      Oregon              9025.067525       No   Premium   
5   HO81102  California              4745.181764       No     Basic   
6   IW72280  California              5035.035257       No     Basic   
7   IH64929  California              4956.247235       No     Basic   
8   FM55990  California              5989.773931      Yes   Premium   
9   QX31376      Oregon             17114.328150       No  Extended   
10  HG93801     Arizona              5154.764074       No  Extended   
11  GE63585  California              2420.711888       No     Basic   
12  EO51920      Oregon             19955.257130       No  Extended   
13  KR82385  California              5454.587929       No     Basic   
14  TL59568     Arizona             12836.009510       No   Premium   
15  CW49887  California              4626.801093      Yes     Basic   
16  FH51383  California              5326.677654       No     Basic   
17  TR26615      Oregon              2497.531509       No     Basic   
18  ZC97968  California              2736.352663       No     Basic   
19  NJ54277  California              3746.751625      Yes  Extended   
20  OM80593  California              7247.313475       No     Basic   
21  RL65736      Oregon              6274.447705       No  Extended   
22  EQ62488      Oregon             15031.564500       No     Basic   
23  NQ71171  California              5107.071054       No     Basic   
24  YK32760      Oregon              2502.209885       No     Basic   
25  AY59197      Nevada              9521.222806       No  Extended   
26  XF15789     Arizona              8869.619888       No     Basic   
27  MQ68407      Oregon              4376.363592      Yes   Premium   
28  CI99444      Nevada              2926.886293       No     Basic   
29  BD70337  California             24706.920110       No     Basic   

               education effective_to_date employmentstatus gender  income  \
0                College        2011-02-18         Employed      M   48029   
1                College        2011-01-18       Unemployed      F       0   
2               Bachelor        2011-02-10         Employed      M   22139   
3                College        2011-01-11         Employed      M   49078   
4               Bachelor        2011-01-17    Medical Leave      F   23675   
5   High School or Below        2011-02-14         Employed      M   50549   
6                 Doctor        2011-02-14         Employed      F   37405   
7   High School or Below        2011-02-24         Employed      M   87197   
8                College        2011-01-19         Employed      M   66839   
9                 Master        2011-01-04         Employed      F   70961   
10  High School or Below        2011-01-02         Employed      M   82730   
11              Bachelor        2011-02-07         Employed      M   31710   
12  High School or Below        2011-01-31         Employed      F   69096   
13                Master        2011-01-26         Employed      M   66334   
14  High School or Below        2011-02-28       Unemployed      M       0   
15                Master        2011-01-16         Employed      F   79487   
16  High School or Below        2011-02-07         Employed      F   76717   
17  High School or Below        2011-01-17       Unemployed      M       0   
18               College        2011-02-14         Disabled      M   17533   
19               College        2011-02-26         Employed      F   41479   
20              Bachelor        2011-02-23       Unemployed      F       0   
21  High School or Below        2011-02-14    Medical Leave      F   13662   
22     

Data Types & Missing Values of Each Column:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10910 entries, 0 to 10909
Data columns (total 26 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   customer                       10910 non-null  object 
 1   state                          10910 non-null  object 
 2   customer_lifetime_value        10910 non-null  float64
 3   response                       10910 non-null  object 
 4   coverage                       10910 non-null  object 
 5   education                      10910 non-null  object 
 6   effective_to_date              10910 non-null  object 
 7   employmentstatus               10910 non-null  object 
 8   gender                         10910 non-null  object 
 9   income                         10910 non-null  int64  
 10  location_code                  10910 non-null  object 
 11  marital_status                 10910 non-null  object 
 12  mo

None

In [1027]:
# check unique values of each column -----------------------------------------------------------------------
# Display the list of column names in df
print("Column Names in df_main:")
display(df.columns.tolist())

# Loop through each column and print unique values
for col in df.columns:
    unique_vals = df[col].unique()
    print(f"Column: {col}")
    print(f"Unique values ({len(unique_vals)}): {unique_vals}\n")

print(f"\nData types check:")
for col in df.columns:
    dtype = df[col].dtype
    unique_vals = df[col].nunique()
    print(f"  ➡️{col:<25} {str(dtype):<10} [{unique_vals}] unique values")

# save the dataset after exploring it into a new CSV file
df.to_csv("marketing_customer_analysis_00_explore.csv", index=False)

Column Names in df_main:


['customer',
 'state',
 'customer_lifetime_value',
 'response',
 'coverage',
 'education',
 'effective_to_date',
 'employmentstatus',
 'gender',
 'income',
 'location_code',
 'marital_status',
 'monthly_premium_auto',
 'months_since_last_claim',
 'months_since_policy_inception',
 'number_of_open_complaints',
 'number_of_policies',
 'policy_type',
 'policy',
 'renew_offer_type',
 'sales_channel',
 'total_claim_amount',
 'vehicle_class',
 'vehicle_size',
 'vehicle_type',
 'month']

Column: customer
Unique values (9134): ['DK49336' 'KX64629' 'LZ68649' ... 'KX53892' 'TL39050' 'WA60547']

Column: state
Unique values (5): ['Arizona' 'California' 'Washington' 'Oregon' 'Nevada']

Column: customer_lifetime_value
Unique values (8041): [ 4809.21696   2228.525238 14947.9173   ...  5259.444853 23893.3041
 11971.97765 ]

Column: response
Unique values (2): ['No' 'Yes']

Column: coverage
Unique values (3): ['Basic' 'Extended' 'Premium']

Column: education
Unique values (5): ['College' 'Bachelor' 'High School or Below' 'Doctor' 'Master']

Column: effective_to_date
Unique values (59): ['2011-02-18' '2011-01-18' '2011-02-10' '2011-01-11' '2011-01-17'
 '2011-02-14' '2011-02-24' '2011-01-19' '2011-01-04' '2011-01-02'
 '2011-02-07' '2011-01-31' '2011-01-26' '2011-02-28' '2011-01-16'
 '2011-02-26' '2011-02-23' '2011-01-15' '2011-02-02' '2011-02-15'
 '2011-01-24' '2011-02-21' '2011-02-22' '2011-01-07' '2011-01-28'
 '2011-02-08' '2011-02-12' '2011-02-20' '2011-01-05' '2011-02-19'
 '20

In [1028]:
# 01 - standardize column names =====================================================================================================

import pandas as pd
# read csv into a new DataFrame
df = pd.read_csv("marketing_customer_analysis_00_explore.csv")

# rename column employmentstatus to employment_status
df.rename(columns={'employmentstatus': 'employment_status'}, inplace=True)

# rename column customer to customer_id
df.rename(columns={'customer': 'customer_id'}, inplace=True)

# save to csv after standardizing column names
df.to_csv("marketing_customer_analysis_01_standardize.csv", index=False)

<hr>

## 2️⃣ INVALID VALUES


<style>
h1 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h2 {
    text-align: left;
    color: darkblue;
    font-weight: bold;
}
</style>

<style>
h3 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h4 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>
<hr>

In [1029]:
# 02 - deal with invalid values in the dataset ======================================================================================

import pandas as pd
# read csv into a new DataFrame
df = pd.read_csv("marketing_customer_analysis_01_standardize.csv")

# number_of_open_complaints --------------------------------------------------------------------------
print("Before: 'number_of_open_complaints' unique values:", df['number_of_open_complaints'].unique())

# round numbers to the nearest integer in number_of_open_complaints column
df['number_of_open_complaints'] = df['number_of_open_complaints'].round().astype('Int64')

# replace 0.38425611  to 0 in number_of_open_complaints column
df['number_of_open_complaints'] = df['number_of_open_complaints'].replace(0.38425611, 0)

# Check unique values after formatting
print("After: 'number_of_open_complaints' unique values:", df['number_of_open_complaints'].unique())
print(*70*"-")


# vehicle_class --------------------------------------------------------------------------
print("Before: 'vehicle_class' unique values:", df['vehicle_class'].unique())
# mapping vehicle_class values
vehicle_mapping = {
    "Sports Car": "Luxury",
    "Luxury SUV": "Luxury",
    "Luxury Car": "Luxury"
}
# replace values in vehicle_class
df['vehicle_class'] = df['vehicle_class'].replace(vehicle_mapping)
# Check unique values after mapping
print("After: 'vehicle_class' unique values:", df['vehicle_class'].unique())
print(*70*"-")


# vehicle_size --------------------------------------------------------------------------
print("Before: 'vehicle_size' unique values:", df['vehicle_size'].unique())
# replace Medsize to Medium in vehicle_size
df['vehicle_size'] = df['vehicle_size'].replace('Medsize', 'Medium')
# Check unique values after mapping
print("After: 'vehicle_size' unique values:", df['vehicle_size'].unique())
print(*70*"-")


# month ---------------------------------------------------------------------------
print("Before: 'month' unique values:", df['month'].unique())
# replace 1 to Jan and 2 to Feb in month column
month_mapping = {
    1: 'Jan',
    2: 'Feb'
}
df['month'] = df['month'].replace(month_mapping)
# Check unique values after mapping
print("After: 'month' unique values:", df['month'].unique())
print(*70*"-")


# display the dataset after dealing with invalid values ----------------------------------------------------------------------
print("Dataset after dealing with invalid values:")
print("Dataset shape:", df.shape)
display(df.head())

# check unique values of each column -----------------------------------------------------------------------
# Display the list of column names in df
print("Column Names in df_main:")
display(df.columns.tolist())

# Loop through each column and print unique values
for col in df.columns:
    unique_vals = df[col].unique()
    print(f"Column: {col}")
    print(f"Unique values ({len(unique_vals)}): {unique_vals}\n")

print(f"\nData types check:")
for col in df.columns:
    dtype = df[col].dtype
    unique_vals = df[col].nunique()
    print(f"  ➡️{col:<25} {str(dtype):<10} [{unique_vals}] unique values")

print(*70*"-")

df.info()
# save step 02 deal with invalid values dataset into a new CSV file ------------------------------
df.to_csv("marketing_customer_analysis_02_invalid-values.csv", index=False)

Before: 'number_of_open_complaints' unique values: [0.         0.38425611 3.         1.         2.         4.
 5.        ]
After: 'number_of_open_complaints' unique values: <IntegerArray>
[0, 3, 1, 2, 4, 5]
Length: 6, dtype: Int64
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Before: 'vehicle_class' unique values: ['Four-Door Car' 'SUV' 'Two-Door Car' 'Sports Car' 'Luxury Car'
 'Luxury SUV']
After: 'vehicle_class' unique values: ['Four-Door Car' 'SUV' 'Two-Door Car' 'Luxury']
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Before: 'vehicle_size' unique values: ['Medsize' 'Small' 'Large']
After: 'vehicle_size' unique values: ['Medium' 'Small' 'Large']
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Before: 'month'

customer_id       state  customer_lifetime_value response  coverage  \
0     DK49336     Arizona              4809.216960       No     Basic   
1     KX64629  California              2228.525238       No     Basic   
2     LZ68649  Washington             14947.917300       No     Basic   
3     XL78013      Oregon             22332.439460      Yes  Extended   
4     QA50777      Oregon              9025.067525       No   Premium   

  education effective_to_date employment_status gender  income  ...  \
0   College        2011-02-18          Employed      M   48029  ...   
1   College        2011-01-18        Unemployed      F       0  ...   
2  Bachelor        2011-02-10          Employed      M   22139  ...   
3   College        2011-01-11          Employed      M   49078  ...   
4  Bachelor        2011-01-17     Medical Leave      F   23675  ...   

  number_of_policies     policy_type        policy  renew_offer_type  \
0                  9  Corporate Auto  Corporate L3            Offer3   
1                  1   Personal Auto   Personal L3            Offer4   
2                  2   Personal Auto   Personal L3            Offer3   
3                  2  Corporate Auto  Corporate L3            Offer2   
4                  7   Personal Auto   Personal L2            Offer1   

   sales_channel  total_claim_amount  vehicle_class vehicle_size vehicle_type  \
0          Agent          292.800000  Four-Door Car       Medium            A   
1    Call Center          744.924331  Four-Door Car       Medium            A   
2    Call Center          480.000000            SUV       Medium            A   
3         Branch          484.013411  Four-Door Car       Medium            A   
4         Branch          707.925645  Four-Door Car       Medium            A   

  month  
0   Feb  
1   Jan  
2   Feb  
3   Jan  
4   Jan  

[5 rows x 26 columns]

Column Names in df_main:


['customer_id',
 'state',
 'customer_lifetime_value',
 'response',
 'coverage',
 'education',
 'effective_to_date',
 'employment_status',
 'gender',
 'income',
 'location_code',
 'marital_status',
 'monthly_premium_auto',
 'months_since_last_claim',
 'months_since_policy_inception',
 'number_of_open_complaints',
 'number_of_policies',
 'policy_type',
 'policy',
 'renew_offer_type',
 'sales_channel',
 'total_claim_amount',
 'vehicle_class',
 'vehicle_size',
 'vehicle_type',
 'month']

Column: customer_id
Unique values (9134): ['DK49336' 'KX64629' 'LZ68649' ... 'KX53892' 'TL39050' 'WA60547']

Column: state
Unique values (5): ['Arizona' 'California' 'Washington' 'Oregon' 'Nevada']

Column: customer_lifetime_value
Unique values (8041): [ 4809.21696   2228.525238 14947.9173   ...  5259.444853 23893.3041
 11971.97765 ]

Column: response
Unique values (2): ['No' 'Yes']

Column: coverage
Unique values (3): ['Basic' 'Extended' 'Premium']

Column: education
Unique values (5): ['College' 'Bachelor' 'High School or Below' 'Doctor' 'Master']

Column: effective_to_date
Unique values (59): ['2011-02-18' '2011-01-18' '2011-02-10' '2011-01-11' '2011-01-17'
 '2011-02-14' '2011-02-24' '2011-01-19' '2011-01-04' '2011-01-02'
 '2011-02-07' '2011-01-31' '2011-01-26' '2011-02-28' '2011-01-16'
 '2011-02-26' '2011-02-23' '2011-01-15' '2011-02-02' '2011-02-15'
 '2011-01-24' '2011-02-21' '2011-02-22' '2011-01-07' '2011-01-28'
 '2011-02-08' '2011-02-12' '2011-02-20' '2011-01-05' '2011-02-19'
 

<hr>

## 3️⃣ DATA TYPES


<style>
h1 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h2 {
    text-align: left;
    color: darkblue;
    font-weight: bold;
}
</style>

<style>
h3 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h4 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>
<hr>

In [1030]:
import pandas as pd

# Load the step 03 dataset into a new DataFrame
df = pd.read_csv("marketing_customer_analysis_02_invalid-values.csv")

# check data types before data formatting
print("Data types before data formatting:")
df.info()

Data types before data formatting:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10910 entries, 0 to 10909
Data columns (total 26 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   customer_id                    10910 non-null  object 
 1   state                          10910 non-null  object 
 2   customer_lifetime_value        10910 non-null  float64
 3   response                       10910 non-null  object 
 4   coverage                       10910 non-null  object 
 5   education                      10910 non-null  object 
 6   effective_to_date              10910 non-null  object 
 7   employment_status              10910 non-null  object 
 8   gender                         10910 non-null  object 
 9   income                         10910 non-null  int64  
 10  location_code                  10910 non-null  object 
 11  marital_status                 10910 non-null  object 
 12  monthly_pre

In [1031]:
# convert effective_to_date column to datetime format
print("Before: 'effective_to_date' data type:", df['effective_to_date'].dtype)
df['effective_to_date'] = pd.to_datetime(df['effective_to_date'], errors='coerce')
print("After: 'effective_to_date' data type:", df['effective_to_date'].dtype)
print("*" * 70)

# convert income to float
print("Before: 'income' data type:", df['income'].dtype)
df['income'] = pd.to_numeric(df['income'], errors='coerce').astype(float)
print("After: 'income' data type:", df['income'].dtype)
print("*" * 70)

# convert months_since_last_claim to integer and round to the nearest integer
print("Before: 'months_since_last_claim' data type:", df['months_since_last_claim'].dtype)
df['months_since_last_claim'] = pd.to_numeric(df['months_since_last_claim'], errors='coerce').round().astype('Int64')
print("After: 'months_since_last_claim' data type:", df['months_since_last_claim'].dtype)
print("*" * 70)

# check data types after data formatting
print("Data types after data formatting:")
df.info()
print("*" * 70)

# display the dataset after data formatting ----------------------------------------------------------------------
print("Dataset after data formatting:")
print("Dataset shape:", df.shape)
display(df.head())

# save step 3 data formatting dataset into a new CSV file ------------------------------
df.to_csv("marketing_customer_analysis_03_data-formatting.csv", index=False)

Before: 'effective_to_date' data type: object
After: 'effective_to_date' data type: datetime64[ns]
**********************************************************************
Before: 'income' data type: int64
After: 'income' data type: float64
**********************************************************************
Before: 'months_since_last_claim' data type: float64
After: 'months_since_last_claim' data type: Int64
**********************************************************************
Data types after data formatting:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10910 entries, 0 to 10909
Data columns (total 26 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   customer_id                    10910 non-null  object        
 1   state                          10910 non-null  object        
 2   customer_lifetime_value        10910 non-null  float64       
 3   response                       10

customer_id       state  customer_lifetime_value response  coverage  \
0     DK49336     Arizona              4809.216960       No     Basic   
1     KX64629  California              2228.525238       No     Basic   
2     LZ68649  Washington             14947.917300       No     Basic   
3     XL78013      Oregon             22332.439460      Yes  Extended   
4     QA50777      Oregon              9025.067525       No   Premium   

  education effective_to_date employment_status gender   income  ...  \
0   College        2011-02-18          Employed      M  48029.0  ...   
1   College        2011-01-18        Unemployed      F      0.0  ...   
2  Bachelor        2011-02-10          Employed      M  22139.0  ...   
3   College        2011-01-11          Employed      M  49078.0  ...   
4  Bachelor        2011-01-17     Medical Leave      F  23675.0  ...   

  number_of_policies     policy_type        policy  renew_offer_type  \
0                  9  Corporate Auto  Corporate L3            Offer3   
1                  1   Personal Auto   Personal L3            Offer4   
2                  2   Personal Auto   Personal L3            Offer3   
3                  2  Corporate Auto  Corporate L3            Offer2   
4                  7   Personal Auto   Personal L2            Offer1   

   sales_channel  total_claim_amount  vehicle_class vehicle_size vehicle_type  \
0          Agent          292.800000  Four-Door Car       Medium            A   
1    Call Center          744.924331  Four-Door Car       Medium            A   
2    Call Center          480.000000            SUV       Medium            A   
3         Branch          484.013411  Four-Door Car       Medium            A   
4         Branch          707.925645  Four-Door Car       Medium            A   

  month  
0   Feb  
1   Jan  
2   Feb  
3   Jan  
4   Jan  

[5 rows x 26 columns]

<hr>

## 4️⃣ NULL VALUES


<style>
h1 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h2 {
    text-align: left;
    color: darkblue;
    font-weight: bold;
}
</style>

<style>
h3 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h4 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>
<hr>

In [1032]:
import pandas as pd
# read csv
df = pd.read_csv("marketing_customer_analysis_03_data-formatting.csv")

# Check for null values
null_counts = df.isnull().sum()
print(f"Total null count in the Dataframe = {null_counts.sum()}\n")

# Identify how many null values each column has. 
print("Columns with null values:")
display(null_counts[null_counts > 0])

print("-" * 70)

# show null values in each column
print("Null values in each column:")
# Loop through each column and print the number of nulls in each column
for col in df.columns:
    null_count = df[col].isnull().sum()
    print(f"Column '{col}'➡️[{null_count}] null values")

# save the dataset after checking for null values into a new CSV file
df.to_csv("marketing_customer_analysis_04_no-nulls.csv", index=False)

Total null count in the Dataframe = 0

Columns with null values:


Series([], dtype: int64)

----------------------------------------------------------------------
Null values in each column:
Column 'customer_id'➡️[0] null values
Column 'state'➡️[0] null values
Column 'customer_lifetime_value'➡️[0] null values
Column 'response'➡️[0] null values
Column 'coverage'➡️[0] null values
Column 'education'➡️[0] null values
Column 'effective_to_date'➡️[0] null values
Column 'employment_status'➡️[0] null values
Column 'gender'➡️[0] null values
Column 'income'➡️[0] null values
Column 'location_code'➡️[0] null values
Column 'marital_status'➡️[0] null values
Column 'monthly_premium_auto'➡️[0] null values
Column 'months_since_last_claim'➡️[0] null values
Column 'months_since_policy_inception'➡️[0] null values
Column 'number_of_open_complaints'➡️[0] null values
Column 'number_of_policies'➡️[0] null values
Column 'policy_type'➡️[0] null values
Column 'policy'➡️[0] null values
Column 'renew_offer_type'➡️[0] null values
Column 'sales_channel'➡️[0] null values
Column 'total_claim_amount'➡️[0] nul

<hr>

## 5️⃣ DUPLICATE VALUES


<style>
h1 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h2 {
    text-align: left;
    color: darkblue;
    font-weight: bold;
}
</style>

<style>
h3 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h4 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>
<hr>

In [1033]:
import pandas as pd

# read csv into a new DataFrame
df = pd.read_csv("marketing_customer_analysis_04_no-nulls.csv")

# Check for duplicate rows in the DataFrame
duplicate_rows = df[df.duplicated()]

# Shape of the dataset before dropping duplicate rows
print("Dataset shape before dropping duplicates:", df.shape)

# Print the number of duplicate rows
print(f"Number of duplicate rows: {duplicate_rows.shape[0]}")

# Display the duplicate rows
if duplicate_rows.empty:
    print("No duplicate rows found.")
else:
    print("\nDuplicate rows:")
    display(duplicate_rows)

print(*70*"-")

# drop duplicate rows, keeping the first occurrence
print("Dropping duplicate rows, keeping the first occurrence...")
df.drop_duplicates(keep='first', inplace=True)
duplicate_rows = df[df.duplicated()]

# Display the duplicate rows
if duplicate_rows.empty:
    print("No duplicate rows found.")
else:
    print("\nDuplicate rows:")
    display(duplicate_rows)
    
# save the dataset after dropping duplicate rows into a new CSV file
df.to_csv("marketing_customer_analysis_05_no-duplicates.csv", index=False)
print(*70*"-")


Dataset shape before dropping duplicates: (10910, 26)
Number of duplicate rows: 460

Duplicate rows:


customer_id       state  customer_lifetime_value response  coverage  \
644       SK12432     Arizona              2388.353882       No     Basic   
645       BZ85280      Oregon              7284.262288       No     Basic   
707       KX23516  California              4455.843406       No  Extended   
895       ON44465  California              5313.294035       No  Extended   
901       YZ19736     Arizona              2580.849899      Yes     Basic   
...           ...         ...                      ...      ...       ...   
10872     XT67997  California              5979.724161      Yes  Extended   
10886     OS84395      Nevada              7611.949286       No  Extended   
10888     MN95006  California              6928.601611       No     Basic   
10894     EX63393  California              6314.492009       No  Extended   
10899     WU60905  California              8610.667489       No     Basic   

                  education effective_to_date employment_status gender  \
644                Bachelor        2011-02-14        Unemployed      M   
645                 College        2011-02-23     Medical Leave      M   
707    High School or Below        2011-02-20        Unemployed      F   
895    High School or Below        2011-01-24        Unemployed      F   
901                 College        2011-02-09          Employed      M   
...                     ...               ...               ...    ...   
10872  High School or Below        2011-02-07          Employed      F   
10886              Bachelor        2011-02-25          Employed      M   
10888              Bachelor        2011-01-10          Employed      M   
10894                Master        2011-02-27          Employed      F   
10899  High School or Below        2011-02-06        Unemployed      F   

        income  ... number_of_policies     policy_type        policy  \
644        0.0  ...                  1   Personal Auto   Personal L1   
645    10334.0  ...                  2   Personal Auto   Personal L1   
707        0.0  ...                  1   Personal Auto   Personal L3   
895        0.0  ...                  9   Personal Auto   Personal L2   
901    46900.0  ...                  1   Personal Auto   Personal L2   
...        ...  ...                ...             ...           ...   
10872  57693.0  ...                  3   Personal Auto   Personal L3   
10886  90252.0  ...                  7  Corporate Auto  Corporate L2   
10888  22907.0  ...                  1   Personal Auto   Personal L2   
10894  66947.0  ...                  6   Personal Auto   Personal L1   
10899      0.0  ...                  9  Corporate Auto  Corporate L1   

       renew_offer_type  sales_channel  total_claim_amount  vehicle_class  \
644              Offer2            Web          677.970934  Four-Door Car   
645              Offer3          Agent          296.962938  Four-Door Car   
707              Offer4          Agent          699.615192            SUV   
895              Offer1          Agent          863.394700  Four-Door Car   
901              Offer2            Web          316.800000  Four-Door Car   
...                 ...            ...                 ...            ...   
10872            Offer2          Agent          547.200000  Four-Door Car   
10886            Offer2            Web          122.924197  Four-Door Car   
10888            Offer4    Call Center           49.187760         Luxury   
10894            Offer3          Agent          374.400000  Four-Door Car   
10899            Offer3            Web          532.800000            SUV   

      vehicle_size vehicle_type month  
644         Medium            A   Feb  
645         Medium            A   Feb  
707         Medium            A   Feb  
895         Medium            A   Jan  
901         Medium            A   Feb  
...            ...          ...   ...  
10872       Medium            A   Feb  
10886       Medium            A   Feb  
10888       Medium            A   Jan  
10

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Dropping duplicate rows, keeping the first occurrence...
No duplicate rows found.
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -


In [1034]:
import pandas as pd
# read csv into a new DataFrame
df = pd.read_csv("marketing_customer_analysis_05_no-duplicates.csv")

# display the dataset ----------------------------------------------------------------------
print("Clean Dataset:")
print("Clean Dataset shape:", df.shape)
display(df.head(20))

# save the clean dataset into a new CSV file
df.to_csv("marketing_customer_analysis_06_clean.csv", index=False)

Clean Dataset:
Clean Dataset shape: (10450, 26)


customer_id       state  customer_lifetime_value response  coverage  \
0      DK49336     Arizona              4809.216960       No     Basic   
1      KX64629  California              2228.525238       No     Basic   
2      LZ68649  Washington             14947.917300       No     Basic   
3      XL78013      Oregon             22332.439460      Yes  Extended   
4      QA50777      Oregon              9025.067525       No   Premium   
5      HO81102  California              4745.181764       No     Basic   
6      IW72280  California              5035.035257       No     Basic   
7      IH64929  California              4956.247235       No     Basic   
8      FM55990  California              5989.773931      Yes   Premium   
9      QX31376      Oregon             17114.328150       No  Extended   
10     HG93801     Arizona              5154.764074       No  Extended   
11     GE63585  California              2420.711888       No     Basic   
12     EO51920      Oregon             19955.257130       No  Extended   
13     KR82385  California              5454.587929       No     Basic   
14     TL59568     Arizona             12836.009510       No   Premium   
15     CW49887  California              4626.801093      Yes     Basic   
16     FH51383  California              5326.677654       No     Basic   
17     TR26615      Oregon              2497.531509       No     Basic   
18     ZC97968  California              2736.352663       No     Basic   
19     NJ54277  California              3746.751625      Yes  Extended   

               education effective_to_date employment_status gender   income  \
0                College        2011-02-18          Employed      M  48029.0   
1                College        2011-01-18        Unemployed      F      0.0   
2               Bachelor        2011-02-10          Employed      M  22139.0   
3                College        2011-01-11          Employed      M  49078.0   
4               Bachelor        2011-01-17     Medical Leave      F  23675.0   
5   High School or Below        2011-02-14          Employed      M  50549.0   
6                 Doctor        2011-02-14          Employed      F  37405.0   
7   High School or Below        2011-02-24          Employed      M  87197.0   
8                College        2011-01-19          Employed      M  66839.0   
9                 Master        2011-01-04          Employed      F  70961.0   
10  High School or Below        2011-01-02          Employed      M  82730.0   
11              Bachelor        2011-02-07          Employed      M  31710.0   
12  High School or Below        2011-01-31          Employed      F  69096.0   
13                Master        2011-01-26          Employed      M  66334.0   
14  High School or Below        2011-02-28        Unemployed      M      0.0   
15                Master        2011-01-16          Employed      F  79487.0   
16  High School or Below        2011-02-07          Employed      F  76717.0   
17  High School or Below        2011-01-17        Unemployed      M      0.0   
18               College        2011-02-14          Disabled      M  17533.0   
19               College        2011-02-26          Employed      F  41479.0   

    ... number_of_policies     policy_type        policy  renew_offer_type  \
0   ...                  9  Corporate Auto  Corporate L3            Offer3   
1   ...                  1   Personal Auto   Personal L3            Offer4   
2   ...                  2   Personal Auto   Personal L3            Offer3   
3   ...                  2  Corporate Auto  Corporate L3            Offer2   
4   ...                  7   Personal Auto   Personal L2            Offer1   
5   ...                  7   Personal Auto   Personal L3            Offer1   
6   ...                  4  Corporate Auto  Corporate L2            Offer2   
7   ...                  3   Personal Auto   Personal L2            Offer2   
8   ...                  1   Personal Auto   Personal L1         

<hr>

## 6️⃣ ANALYSIS


<style>
h1 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h2 {
    text-align: left;
    color: darkblue;
    font-weight: bold;
}
</style>

<style>
h3 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>

<style>
h4 {
    text-align: center;
    color: black;
    font-weight: bold;
}
</style>
<hr>

In [1]:
import pandas as pd

# load the clean dataset into a new DataFrame
df = pd.read_csv("marketing_customer_analysis_06_clean.csv")

# display the dataset
print("Exploring the Clean Marketing Customer Dataset:")

# Display df DataFrame SHAPE
print(f"Clean Dataset Shape: [{df.shape[0]}] ROWS and [{df.shape[1]}] COLUMNS\n")

# Display df DataFrame HEAD
display(df.head())

# Display data types & missing values of each column in df DataFrame
print("Data Types & Missing Values of Each Column:")
display(df.info())


Exploring the Clean Marketing Customer Dataset:
Clean Dataset Shape: [10450] ROWS and [26] COLUMNS



customer_id       state  customer_lifetime_value response  coverage  \
0     DK49336     Arizona              4809.216960       No     Basic   
1     KX64629  California              2228.525238       No     Basic   
2     LZ68649  Washington             14947.917300       No     Basic   
3     XL78013      Oregon             22332.439460      Yes  Extended   
4     QA50777      Oregon              9025.067525       No   Premium   

  education effective_to_date employment_status gender   income  ...  \
0   College        2011-02-18          Employed      M  48029.0  ...   
1   College        2011-01-18        Unemployed      F      0.0  ...   
2  Bachelor        2011-02-10          Employed      M  22139.0  ...   
3   College        2011-01-11          Employed      M  49078.0  ...   
4  Bachelor        2011-01-17     Medical Leave      F  23675.0  ...   

  number_of_policies     policy_type        policy  renew_offer_type  \
0                  9  Corporate Auto  Corporate L3            Offer3   
1                  1   Personal Auto   Personal L3            Offer4   
2                  2   Personal Auto   Personal L3            Offer3   
3                  2  Corporate Auto  Corporate L3            Offer2   
4                  7   Personal Auto   Personal L2            Offer1   

   sales_channel  total_claim_amount  vehicle_class vehicle_size vehicle_type  \
0          Agent          292.800000  Four-Door Car       Medium            A   
1    Call Center          744.924331  Four-Door Car       Medium            A   
2    Call Center          480.000000            SUV       Medium            A   
3         Branch          484.013411  Four-Door Car       Medium            A   
4         Branch          707.925645  Four-Door Car       Medium            A   

  month  
0   Feb  
1   Jan  
2   Feb  
3   Jan  
4   Jan  

[5 rows x 26 columns]

Data Types & Missing Values of Each Column:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10450 entries, 0 to 10449
Data columns (total 26 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   customer_id                    10450 non-null  object 
 1   state                          10450 non-null  object 
 2   customer_lifetime_value        10450 non-null  float64
 3   response                       10450 non-null  object 
 4   coverage                       10450 non-null  object 
 5   education                      10450 non-null  object 
 6   effective_to_date              10450 non-null  object 
 7   employment_status              10450 non-null  object 
 8   gender                         10450 non-null  object 
 9   income                         10450 non-null  float64
 10  location_code                  10450 non-null  object 
 11  marital_status                 10450 non-null  object 
 12  mo

None

1. You work at the marketing department and you want to know which sales channel brought the most sales in terms of total revenue. 
Using pivot, create a summary table showing the total revenue for each sales channel (branch, call center, web, and mail).
Round the total revenue to 2 decimal points.  Analyze the resulting table to draw insights.

In [2]:
pivot_table = (
    df.pivot_table(
        index='sales_channel',
        values='customer_lifetime_value',
        aggfunc='sum'
    )
    .round(2)
    .reset_index()
)

pivot_table


sales_channel  customer_lifetime_value
0         Agent              31723780.08
1        Branch              23507902.32
2   Call Center              16523198.75
3           Web              12106814.99

---
Insight:
- Top revenue channel: Agent (~31.7M) = high-value clients, effective upselling.
- Second: Branch (~23.5M) = in-person sales still strong.
- Lower: Call Center (~16.5M) = moderate performance.
- Lowest: Web (~12.1M) = underperforming; opportunity to improve digital sales.

Business Impact: 
- Focus on Agents and Branches for revenue, optimize Call Center, and invest in improving Web channel for scalability.
---

2. Create a pivot table that shows the **average customer lifetime value per gender and education level**. Analyze the resulting table to draw insights.

In [3]:
# Pivot table: average customer lifetime value per gender
pivot_avg_clv_gender = df.pivot_table(
    index='gender',
    values='customer_lifetime_value',
    aggfunc='mean'
).round(2)

pivot_avg_clv_gender = pivot_avg_clv_gender.reset_index()
pivot_avg_clv_gender


gender  customer_lifetime_value
0      F                  8077.64
1      M                  7970.14

In [4]:
df['education'].value_counts()

education
Bachelor                3121
College                 3053
High School or Below    3022
Master                   865
Doctor                   389
Name: count, dtype: int64

In [5]:
df['education'].describe()

count        10450
unique           5
top       Bachelor
freq          3121
Name: education, dtype: object

In [ ]:
# Define logical order for education
education_order = ['High School', 'College', 'Master', 'Doctor', 'PhD']
df['education'] = pd.Categorical(df['education'], categories=education_order, ordered=True)

# Nested pivot table: education as columns, gender as rows
pivot_gender_education = df.pivot_table(
    index='gender',
    columns='education',
    values='customer_lifetime_value',
    aggfunc='mean'
).round(2)  # round values

# Optional: reset columns if you want a flat table
pivot_gender_education = pivot_gender_education.reset_index()

pivot_gender_education


education gender  College   Master   Doctor
0              F  7757.32  8224.62  7369.59
1              M  8075.26  8154.61  7414.86

---
Insight:
- Highest CLV: Females with Master degrees (8224.62).
- Lowest CLV: Females with Doctor degrees (7369.59).
- Males: Highest CLV in College (8075.26), lowest in Doctor (7414.86).
- Education level impacts CLV more than gender; Master and College customers are most valuable.

Business Impact:
- Focus marketing on Master-educated females and College-educated males; consider re-engaging Doctor-level customers.

---
## Bonus

You work at the customer service department and you want to know which months had the highest number of complaints by policy type category. Create a summary table showing the number of complaints by policy type and month.
Show it in a long format table.

*In data analysis, a long format table is a way of structuring data in which each observation or measurement is stored in a separate row of the table. The key characteristic of a long format table is that each column represents a single variable, and each row represents a single observation of that variable.*

*More information about long and wide format tables here: https://www.statology.org/long-vs-wide-data/*

In [1036]:
# Your code goes here